mlir/lib/IR/AttributeDetail.h

```diff
@@ -411,12 +414,56 @@ class DistinctAttributeAllocator {
   /// Allocates a distinct attribute storage using a thread local bump pointer
   /// allocator to enable synchronization free parallel allocations.
   DistinctAttrStorage *allocate(Attribute referencedAttr) {
-    return new (allocatorCache.get().Allocate<DistinctAttrStorage>())
-        DistinctAttrStorage(referencedAttr);
+    if (useSynchronizedAllocator && !useThreadLocalAllocator) {
+      std::scoped_lock<std::mutex> lock(allocatorMutex);
+      return allocateImpl(referencedAttr);
+    }
+    if (!useSynchronizedAllocator)
+      return allocateImpl(referencedAttr);
+    llvm_unreachable(
+        "Cannot use both a synchronised and thread_local allocator!");
```

Comment:
Sorry somehow missed the unreachable before. I think we should to avoid this since there should always be a safe fallback.

Couldn't we write:
```
if (!useThreadLocalAllocator && enableMultiThreading) {
  // Lock the allocator since multi-threading is enabled and the thread local allocators cannot be used.
  std::scoped_lock<std::mutex> lock(allocatorMutex);
  return allocateImpl(referencedAttr);
}
return allocateImpl(referencedAttr);
```